### ISSUE TITLE
Request: Please make FIPS 140-2 compliant images available

### ISSUE THREAD

**Issue by gesarki** [miekg/coredns#6600]
**What would you like to be added**:
FIPS 140-2 compliant images.

**Why is this needed**:
At Acquia we’re currently using this component as part of our globally distributed Kubernetes infrastructure.

Like most providers, we expend a lot of effort ensuring that we remain compliant with varied regulatory regimes across different localities. Currently we're working towards FIPS compliance which is required by federal institutions like healthcare, banking, and defense organizations to set cryptographic standards for highly sensitive data.

The FIPS 140-2 encryption compliance requirement poses a common problem that is likely to be shared by any organizations looking to obtain a FEDRAMP certification and which use a myriad of OSS Kubernetes components under the hood.

FIPS compliance stands as a barrier to entry for smaller organizations. Accepting our contribution would take a small step towards lowering that barrier for the general public and other open source projects.

We, and any other cloud based software providers that are building atop Kubernetes, would benefit from the availability of off-the-shelf FIPS-compliant variants of this project’s images.

While many tools exist to tweak deployment manifests, it can be complex and fragile to rebuild upstream images with constraints placed on encryption libraries. This type of change would be far easier to make inside the originating repo.

Internally, we have explored several paths forward to achieve our compliance objectives. We are hoping to be able to contribute the required changes upstream in a way that benefits the community at large and minimizes toil.

Initially, we are requesting that the current maintainers and community consider supporting FIPS images. We have a notional approach to contribute and would like to collaborate.

Please let us know your thoughts in this Issue or by reaching out directly.

At this time, we have an internal approach that we are using on our internal Go-based controllers and other container images. There are some basic requirements that we’d ask to be built into a `-fips` variant including assertions that could live in the upstream build steps or in later CI steps.

What we do internally is:

- Image is built with a FIPS library such as boringcrypto for Go
- Image is based on Amazon Linux 2, Canonical Ubuntu 20.04 or later, or UBI8
- Image has appropriate group and user specified and final image executes in that context
- We assert that binaries call out to boringcrypto via dynamic linking (this is recommended during build but is not required, we will verify this ourselves in our CI/CD infra)

Thank you for your time considering this request as well as your efforts supporting this project!

Acquia KaaS Core Services team

**Comment by SuperQ** [comment]
> Image is based on Amazon Linux 2, Canonical Ubuntu 20.04 or later, or UBI8

We typically recommend going with a distroless container image for high security environments. CoreDNS is statically compiled and requires no OS libraries.

> We assert that binaries call out to boringcrypto via dynamic linking (this is recommended during build but is not required, we will verify this ourselves in our CI/CD infra)

Dynamic linking is not something we would want to support. We prefer pure go libraries.

**Comment by max-allan-cgr** [comment]
@gesarki , you could try Chainguard's image: https://images.chainguard.dev/directory/image/coredns-fips/versions

When Coredns version 1.11.2 (or 1.11.3) is tagged and released, there will be a version that is free of known vulnerabilities.

(Apologies for advertising, I do work for Chainguard. Non-fips "latest" images are free, but this (and other fips'ed) images are not, they take a lot of effort to create and maintain.)

**Comment by imuni4fun** [comment]
@SuperQ ah, thank you! I, too, am heavily biased towards `scratch` or the less extreme option `distroless`, both for security reasons and for sheer size/distribution reasons. It's hard to attack a container when there's almost nothing in it 👍 

I'm on @gesarki's team at Acquia and we missed that this was statically compiled. We'll have to circle back on our approach for this and then come back to this issue/thread.

@max-allan-cgr no issues for the advertisement 😄. We agree that compliance can be quite the burden. We're trying to reduce that burden for ourselves and for others via this proposed change.

**Comment by SuperQ** [comment]
Yes, I don't think we will be able to support any solutions here that require `CGO_ENABLED=1`.

Using linked C libraries is extremely problematic for production use. It is dangerous and fragile. For example, at my `$dayjob`, we are just in the process of banning C library linking from Go in our production environment without specific exceptions. And those exceptions come with the expectation that teams will re-write the C code into Go long-term.

As a member of the CoreDNS steering committee, my vote is that for any solution for FIPS 140-2 compliance in CoreDNS must be native Go code.
